File: App/app/src/main/java/com/cinema/ticket_booking/data/repository/MovieRepository.java
package com.cinema.ticket_booking.data.repository;

import androidx.lifecycle.*;
import com.cinema.ticket_booking.data.model.response.*;
import com.cinema.ticket_booking.network.ApiService;
import com.cinema.ticket_booking.util.Resource;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.List;

@Singleton
public class MovieRepository {

    private final ApiService apiService;

    @Inject
    public MovieRepository(ApiService apiService) {
        this.apiService = apiService;
    }

    public LiveData<Resource<PageResponse<MovieSummary>>> getMovies(String status, int page, int size) {
        MutableLiveData<Resource<PageResponse<MovieSummary>>> result = new MutableLiveData<>();
        result.setValue(Resource.loading());
        apiService.getMovies(status, page, size).enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<PageResponse<MovieSummary>>> call,
                    Response<ApiResponse<PageResponse<MovieSummary>>> response) {
                if (response.isSuccessful() && response.body() != null && response.body().isSuccess())
                    result.setValue(Resource.success(response.body().getData()));
                else
                    result.setValue(Resource.error("Tải phim thất bại"));
            }

            @Override
            public void onFailure(Call<ApiResponse<PageResponse<MovieSummary>>> call, Throwable t) {
                result.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return result;
    }

    public LiveData<Resource<MovieDetail>> getMovieDetail(String id) {
        MutableLiveData<Resource<MovieDetail>> result = new MutableLiveData<>();
        result.setValue(Resource.loading());
        apiService.getMovieDetail(id).enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<MovieDetail>> call,
                    Response<ApiResponse<MovieDetail>> response) {
                if (response.isSuccessful() && response.body() != null && response.body().isSuccess())
                    result.setValue(Resource.success(response.body().getData()));
                else
                    result.setValue(Resource.error("Không tìm thấy phim"));
            }

            @Override
            public void onFailure(Call<ApiResponse<MovieDetail>> call, Throwable t) {
                result.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return result;
    }

    public LiveData<Resource<CanReviewResponse>> canReview(String id) {
        MutableLiveData<Resource<CanReviewResponse>> result = new MutableLiveData<>();
        result.setValue(Resource.loading());
        apiService.canReview(id).enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<CanReviewResponse>> call,
                    Response<ApiResponse<CanReviewResponse>> response) {
                if (response.isSuccessful() && response.body() != null && response.body().isSuccess())
                    result.setValue(Resource.success(response.body().getData()));
                else
                    result.setValue(Resource.error(response.body() != null ? response.body().getMessage() : "Lỗi kiểm tra quyền đánh giá"));
            }

            @Override
            public void onFailure(Call<ApiResponse<CanReviewResponse>> call, Throwable t) {
                result.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return result;
    }

    public LiveData<Resource<PageResponse<MovieSummary>>> searchMovies(String query, int page, int size) {
        MutableLiveData<Resource<PageResponse<MovieSummary>>> result = new MutableLiveData<>();
        result.setValue(Resource.loading());
        apiService.searchMovies(query, page, size).enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<PageResponse<MovieSummary>>> call,
                    Response<ApiResponse<PageResponse<MovieSummary>>> response) {
                if (response.isSuccessful() && response.body() != null && response.body().isSuccess())
                    result.setValue(Resource.success(response.body().getData()));
                else
                    result.setValue(Resource.error("Tìm kiếm thất bại"));
            }

            @Override
            public void onFailure(Call<ApiResponse<PageResponse<MovieSummary>>> call, Throwable t) {
                result.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return result;
    }

    public LiveData<Resource<List<Genre>>> getGenres() {
        MutableLiveData<Resource<List<Genre>>> result = new MutableLiveData<>();
        result.setValue(Resource.loading());
        apiService.getAllGenres().enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<List<Genre>>> call,
                    Response<ApiResponse<List<Genre>>> response) {
                if (response.isSuccessful() && response.body() != null && response.body().isSuccess())
                    result.setValue(Resource.success(response.body().getData()));
                else
                    result.setValue(Resource.error("Tải thể loại thất bại"));
            }

            @Override
            public void onFailure(Call<ApiResponse<List<Genre>>> call, Throwable t) {
                result.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return result;
    }

    public LiveData<Resource<List<MovieSummary>>> getFeaturedMovies() {
        MutableLiveData<Resource<List<MovieSummary>>> result = new MutableLiveData<>();
        result.setValue(Resource.loading());
        apiService.getFeaturedMovies().enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<List<MovieSummary>>> call,
                    Response<ApiResponse<List<MovieSummary>>> response) {
                if (response.isSuccessful() && response.body() != null && response.body().isSuccess())
                    result.setValue(Resource.success(response.body().getData()));
                else
                    result.setValue(Resource.error("Tải phim nổi bật thất bại"));
            }

            @Override
            public void onFailure(Call<ApiResponse<List<MovieSummary>>> call, Throwable t) {
                result.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return result;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/data/repository/MovieRepository.kt
package com.cinema.ticket_booking.data.repository

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.cinema.ticket_booking.data.model.response.*
import com.cinema.ticket_booking.network.ApiService
import com.cinema.ticket_booking.util.Resource
import javax.inject.Inject
import javax.inject.Singleton
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

@Singleton
class MovieRepository @Inject constructor(
    private val apiService: ApiService
) {

    fun getMovies(status: String, page: Int, size: Int): LiveData<Resource<PageResponse<MovieSummary>>> {
        val result = MutableLiveData<Resource<PageResponse<MovieSummary>>>()
        result.value = Resource.loading()
        apiService.getMovies(status, page, size).enqueue(object : Callback<ApiResponse<PageResponse<MovieSummary>>> {
            override fun onResponse(
                call: Call<ApiResponse<PageResponse<MovieSummary>>>,
                response: Response<ApiResponse<PageResponse<MovieSummary>>>
            ) {
                if (response.isSuccessful && response.body() != null && response.body()!!.success) {
                    result.value = Resource.success(response.body()!!.data)
                } else {
                    result.value = Resource.error("Tải phim thất bại")
                }
            }

            override fun onFailure(call: Call<ApiResponse<PageResponse<MovieSummary>>>, t: Throwable) {
                result.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return result
    }

    fun getMovieDetail(id: String): LiveData<Resource<MovieDetail>> {
        val result = MutableLiveData<Resource<MovieDetail>>()
        result.value = Resource.loading()
        apiService.getMovieDetail(id).enqueue(object : Callback<ApiResponse<MovieDetail>> {
            override fun onResponse(call: Call<ApiResponse<MovieDetail>>, response: Response<ApiResponse<MovieDetail>>) {
                if (response.isSuccessful && response.body() != null && response.body()!!.success) {
                    result.value = Resource.success(response.body()!!.data)
                } else {
                    result.value = Resource.error("Không tìm thấy phim")
                }
            }

            override fun onFailure(call: Call<ApiResponse<MovieDetail>>, t: Throwable) {
                result.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return result
    }

    fun canReview(id: String): LiveData<Resource<CanReviewResponse>> {
        val result = MutableLiveData<Resource<CanReviewResponse>>()
        result.value = Resource.loading()
        apiService.canReview(id).enqueue(object : Callback<ApiResponse<CanReviewResponse>> {
            override fun onResponse(
                call: Call<ApiResponse<CanReviewResponse>>,
                response: Response<ApiResponse<CanReviewResponse>>
            ) {
                if (response.isSuccessful && response.body() != null && response.body()!!.success) {
                    result.value = Resource.success(response.body()!!.data)
                } else {
                    result.value = Resource.error(
                        if (response.body() != null) response.body()!!.message else "Lỗi kiểm tra quyền đánh giá"
                    )
                }
            }

            override fun onFailure(call: Call<ApiResponse<CanReviewResponse>>, t: Throwable) {
                result.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return result
    }

    fun searchMovies(query: String, page: Int, size: Int): LiveData<Resource<PageResponse<MovieSummary>>> {
        val result = MutableLiveData<Resource<PageResponse<MovieSummary>>>()
        result.value = Resource.loading()
        apiService.searchMovies(query, page, size).enqueue(object : Callback<ApiResponse<PageResponse<MovieSummary>>> {
            override fun onResponse(
                call: Call<ApiResponse<PageResponse<MovieSummary>>>,
                response: Response<ApiResponse<PageResponse<MovieSummary>>>
            ) {
                if (response.isSuccessful && response.body() != null && response.body()!!.success) {
                    result.value = Resource.success(response.body()!!.data)
                } else {
                    result.value = Resource.error("Tìm kiếm thất bại")
                }
            }

            override fun onFailure(call: Call<ApiResponse<PageResponse<MovieSummary>>>, t: Throwable) {
                result.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return result
    }

    fun getGenres(): LiveData<Resource<List<Genre>>> {
        val result = MutableLiveData<Resource<List<Genre>>>()
        result.value = Resource.loading()
        apiService.getAllGenres().enqueue(object : Callback<ApiResponse<List<Genre>>> {
            override fun onResponse(call: Call<ApiResponse<List<Genre>>>, response: Response<ApiResponse<List<Genre>>>) {
                if (response.isSuccessful && response.body() != null && response.body()!!.success) {
                    result.value = Resource.success(response.body()!!.data)
                } else {
                    result.value = Resource.error("Tải thể loại thất bại")
                }
            }

            override fun onFailure(call: Call<ApiResponse<List<Genre>>>, t: Throwable) {
                result.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return result
    }

    fun getFeaturedMovies(): LiveData<Resource<List<MovieSummary>>> {
        val result = MutableLiveData<Resource<List<MovieSummary>>>()
        result.value = Resource.loading()
        apiService.getFeaturedMovies().enqueue(object : Callback<ApiResponse<List<MovieSummary>>> {
            override fun onResponse(
                call: Call<ApiResponse<List<MovieSummary>>>,
                response: Response<ApiResponse<List<MovieSummary>>>
            ) {
                if (response.isSuccessful && response.body() != null && response.body()!!.success) {
                    result.value = Resource.success(response.body()!!.data)
                } else {
                    result.value = Resource.error("Tải phim nổi bật thất bại")
                }
            }

            override fun onFailure(call: Call<ApiResponse<List<MovieSummary>>>, t: Throwable) {
                result.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return result
    }
}
